sql: track fine-grained view dependencies

RichardJCai opened the pull request:
Fixes #29021
Only create view dependencies on columns if the column is referenced in the view query.

Note: this does not fix the dependencies of old views.

Release note (sql change): Views now only create a dependency on a table's column if the
column is referenced in the view definition. Previously, all columns were added as a dependency
meaning if a table was referenced in a view, all columns regardless of if the column was actually
referenced would be added to the view's dependencies.

RichardJCai committed:
sql: track fine-grained view dependencies
Fixes #29021
Only create view dependencies on columns if the column is referenced in the view query.

Release note (sql change): Views now only create a dependency on a table's column if the
column is referenced in the view definition. Previously, all columns were added as a dependency
meaning if a table was referenced in a view, all columns regardless of if the column was actually
referenced would be added to the view's dependencies.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49872)
<!-- Reviewable:end -->

RichardJCai commented:
Please let me know if I missed any column references where view dependencies should be added

RaduBerinde reviewed (commented):
I am not entirely convinced the logic covers all cases; even if it does, it seems fragile. We should add a test (similar to TestBuilder) which adds a "view-deps" directive which builds a query and just prints out the view deps. That would allow us to much more easily test all sorts of cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-:-M9-AhA47RyAA19_dDcx:b-2qceg8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/sql/opt/view_dependencies.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A7llEAhbSBmX1Fbk:-M9-A7llEAhbSBmX1Fbl:bg2f4rz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/view_dependencies.go#L31)):*
> ```Go
> 	ColumnOrdinals util.FastIntSet
> 
> 	ColumnIDToOrd map[ColumnID]int
> ```

Any ColumnID that is coming from a TableID can be converted to a table ordinal using TableID.ColumnOrdinal, can we use that instead of adding this map?


<!-- Sent from Reviewable.io -->

rytaft reviewed (changes requested):
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/views, line 697 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-9DsnCr1H-Y1jGZq8:-M9-9DsnCr1H-Y1jGZq9:bwodv72) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/logictest/testdata/logic_test/views#L697)):*
> ```
> # Should be able to drop columns views don't depend on.
> statement ok
> CREATE TABLE tv8 (x INT, y INT);
> ```

[nit] did we already have tv1-tv7? I don't see them above...
___
*[pkg/sql/logictest/testdata/logic_test/views, line 707 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-D-qC0-4l-MT84TkJ:-M9-D-qC0-4l-MT84TkK:b-dvzqwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/logictest/testdata/logic_test/views#L707)):*
> ```
> statement ok
> CREATE TABLE tv9(x INT, y INT);
> CREATE VIEW v9 AS SELECT x FROM tv9 GROUP BY x,y
> ```

you should add some tests where you are also projecting an expression, such as "SELECT x+1 ..."
___
*[pkg/sql/logictest/testdata/logic_test/views, line 728 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-9bw91aP4RCWnohHX:-M9-9bwA6QBD7638Cnbg:b-oqiy7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/logictest/testdata/logic_test/views#L728)):*
> ```
> 
> statement error pq: cannot drop column \"y\" because view \"v10\" depends on it
> ALTER TABLE tv10a DROP COLUMN y
> ```

tv10a -> tv10b

(would help if the table names were a bit more friendly to make this sort of duplicate test easier to spot)
___
*[pkg/sql/logictest/testdata/logic_test/views, line 778 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-ANNr3zNEpeIx8n1H:-M9-ANNr3zNEpeIx8n1I:b-hkg9yq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/logictest/testdata/logic_test/views#L778)):*
> ```
> ALTER TABLE tv13 DROP COLUMN y
> 
> # Dependencies should be tracked in window functions.
> ```

[nit] aggregate and window functions

(your first example below is not a window function)
___
*[pkg/sql/logictest/testdata/logic_test/views, line 789 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A_2VE0-pQymFPqDD:-M9-A_2VE0-pQymFPqDE:b-z7duuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/logictest/testdata/logic_test/views#L789)):*
> ```
> statement ok
> CREATE TABLE tv15(x INT, y INT);
> CREATE VIEW v15 AS SELECT sum(x) OVER (PARTITION by y) FROM tv15;
> ```

what about adding ORDER BY with a different column to the partition too?
___
*[pkg/sql/opt/optbuilder/builder.go, line 342 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-BJ1rExjaaPGTRVf_:-M9-BJ1s1gTzPIxGNbUN:b-1p0egw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/optbuilder/builder.go#L342)):*
> ```Go
> }
> 
> // TrackReferencedColumnForViews is used to add a column to a views dependencies.
> ```

[nit] a views dependencies
___
*[pkg/sql/opt/optbuilder/builder.go, line 348 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-Btyx6rl5exE90Ua-:-M9-Btyy0jJTS4T27dLV:bfafns1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/optbuilder/builder.go#L348)):*
> ```Go
> 		for i := range b.viewDeps {
> 			dep := b.viewDeps[i]
> 			for id, ord := range dep.ColumnIDToOrd {
> ```

Since this is a map you shouldn't need to iterate through the whole thing. Can't you just do:
```
if ord, ok := dep.ColumnIDToOrd[col.id]; ok {
  dep.ColumnOrdinals.Add(ord)
}
```


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> I am not entirely convinced the logic covers all cases; even if it does, it seems fragile. We should add a test (similar to TestBuilder) which adds a "view-deps" directive which builds a query and just prints out the view deps. That would allow us to much more easily test all sorts of cases.
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-:-M9-AhA47RyAA19_dDcx:b-2qceg8)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)
> 
> _[pkg/sql/opt/view_dependencies.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A7llEAhbSBmX1Fbk:-M9-A7llEAhbSBmX1Fbl:bg2f4rz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/view_dependencies.go#L31)):_
> 
> > ```go
> > 	ColumnOrdinals util.FastIntSet
> > 
> > 	ColumnIDToOrd map[ColumnID]int
> > ```
> 
> Any ColumnID that is coming from a TableID can be converted to a table ordinal using TableID.ColumnOrdinal, can we use that instead of adding this map?

I agree that this seems fragile, do you have any suggestions to make it less so? 
Also I like the idea of having a view-deps test directive. Although we still can't exhaustively test all the cases, what kind of testing would we need to be confident?

RaduBerinde reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/sql/opt/view_dependencies.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A7llEAhbSBmX1Fbk:-M9-IE5v2NnNs1755pwL:b752kcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/view_dependencies.go#L31)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I agree that this seems fragile, do you have any suggestions to make it less so?  Also I like the idea of having a view-deps test directive. Although we still can't exhaustively test all the cases, what kind of testing would we need to be confident? 
</blockquote></details>

Not really, other than the test. Just being able to see all the deps at once instead of having to probe specific columns by trying to remove them would make a big difference. Some conditions that we should test in addition to what you already tested:
 - cases where the same table is used multiple times (with different column sets)
 - cases where a column is only used by an ORDER BY (eg `SELECT x FROM t ORDER BY y`), including in ORDER BY inside aggregates
 - cases with an ordered-set aggregate function (eg `SELECT percentile_cont(0.50) WITHIN GROUP (ORDER BY c) FROM t`)


<!-- Sent from Reviewable.io -->

rytaft commented:
I'm also wondering if `optbuilder` is still the right place for this. If we don't want to include columns that are pruned or optimized away in the view dependencies, then maybe we should collect them *after* optimization, inside the execbuilder. What do you think , @raduberinde?





<!-- Sent from Reviewable.io -->

RaduBerinde commented:
We need to include all columns referenced in the query, not just what ends up being used after optimization. If we don't, we can remove columns and then the view query won't work anymore. That is #17269, see an example in there.

rytaft commented:
Ah ok got it -- never mind then!





<!-- Sent from Reviewable.io -->

RichardJCai commented:
> I am not entirely convinced the logic covers all cases; even if it does, it seems fragile. We should add a test (similar to TestBuilder) which adds a "view-deps" directive which builds a query and just prints out the view deps. That would allow us to much more easily test all sorts of cases.
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-:-M9-AhA47RyAA19_dDcx:b-2qceg8)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)
> 
> _[pkg/sql/opt/view_dependencies.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A7llEAhbSBmX1Fbk:-M9-A7llEAhbSBmX1Fbl:bg2f4rz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/view_dependencies.go#L31)):_
> 
> > ```go
> > 	ColumnOrdinals util.FastIntSet
> > 
> > 	ColumnIDToOrd map[ColumnID]int
> > ```
> 
> Any ColumnID that is coming from a TableID can be converted to a table ordinal using TableID.ColumnOrdinal, can we use that instead of adding this map?

I started tracking column names in view deps so they could be printed out using TestBuilder and moved the tests over to use TestBuilder to show the columns, let me know what you think about this, I was also unclear on how to remove the map and use TableIDs.

Also addressed Rebecca's comments and added the suggested tests. 

@RaduBerinde, @rytaft

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rytaft)

___
*[pkg/sql/opt/view_dependencies.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A7llEAhbSBmX1Fbk:-M94WKBcB-cGIHWbxsfg:b-lfo4rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/view_dependencies.go#L31)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I started tracking column names in view deps so they could be printed out using TestBuilder and moved the tests over to use TestBuilder to show the columns, let me know what you think about this, I was also unclear on how to remove the map and use TableIDs.

Also addressed Rebecca's comments and added the suggested tests. 

@RaduBerinde, @rytaft  
</blockquote></details>

I see, we don't have a TableID handy. It's fine to keep the map, just add a comment explaining what it is (and make it clear that it holds all columns of the data source, not just the used ones).

We should be able to get the names using `DataSource.(cat.Table).Column(ord).ColName()` or `DataSource.(cat.View).ColumnName(ord)` (this could be a ViewDeps method) instead of storing them separately.
___
*[pkg/sql/opt/view_dependencies.go, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94WYGG8QYis2a3FCvZ:-M94WYGG8QYis2a3FCv_:bwof1af) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/view_dependencies.go#L27)):*
> ```Go
> 
> 	// ColumnOrdinals is the set of column ordinals that are referenced by the
> 	// view for this table. In most cases, this consists of all "public" columns
> ```

This comment needs updating.
___
*[pkg/sql/opt/memo/expr_format.go, line 556 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94XRUS4Yb9DJEAjhQE:-M94XRUS4Yb9DJEAjhQF:bemvvpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/memo/expr_format.go#L556)):*
> ```Go
> 			}
> 			if !dep.ColumnOrdinals.Empty() {
> 				fmt.Fprintf(f.Buffer, " [column ordinals: %s]", dep.ColumnOrdinals)
> ```

I don't think we need to print both the ordinals and the names. If we have names, we should use them.
___
*[pkg/sql/opt/optbuilder/builder.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94XFD21ryUtlhNuUsd:-M94XFD21ryUtlhNuUse:ben42d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/builder.go#L342)):*
> ```Go
> }
> 
> // TrackReferencedColumnForViews is used to add a column to a views dependencies.
> ```

[nit] to the view dependencies
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94Xu-v44Lg5YQGZ4cT:-M94Xu-v44Lg5YQGZ4cU:b249avz) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/testdata/create_view#L93)):*
> ```
>  └── dependencies
>       ├── ab [column ordinals: (0)] [column names: [a]]
>       ├── ab
> ```

does this mean we don't depend on any columns? maybe change the formatter to print that explicitly somehow (e.g. `ab [no columns]`).
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94Y9PuCksSLQsuYfnU:-M94Y9Pv5LgbUBI__YdT:b-sk34fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/testdata/create_view#L106)):*
> ```
> error (42601): CREATE VIEW specifies 1 column name, but data source has 2 columns
> 
> build
> ```

Why did you remove this comment?


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rytaft)
> 
> _[pkg/sql/opt/view_dependencies.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9-A7llEAhbSBmX1Fbk:-M94WKBcB-cGIHWbxsfg:b-lfo4rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0871edafb699a787de4c963f806c513eaed51a/pkg/sql/opt/view_dependencies.go#L31)):_
> 
> Previously, RichardJCai (Richard Cai) wrote…
> I see, we don't have a TableID handy. It's fine to keep the map, just add a comment explaining what it is (and make it clear that it holds all columns of the data source, not just the used ones).
> 
> We should be able to get the names using `DataSource.(cat.Table).Column(ord).ColName()` or `DataSource.(cat.View).ColumnName(ord)` (this could be a ViewDeps method) instead of storing them separately.
> 
> _[pkg/sql/opt/view_dependencies.go, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94WYGG8QYis2a3FCvZ:-M94WYGG8QYis2a3FCv_:bwof1af) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/view_dependencies.go#L27)):_
> 
> > ```go
> > 	// ColumnOrdinals is the set of column ordinals that are referenced by the
> > 	// view for this table. In most cases, this consists of all "public" columns
> > ```
> 
> This comment needs updating.
> 
> _[pkg/sql/opt/memo/expr_format.go, line 556 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94XRUS4Yb9DJEAjhQE:-M94XRUS4Yb9DJEAjhQF:bemvvpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/memo/expr_format.go#L556)):_
> 
> > ```go
> > 			}
> > 			if !dep.ColumnOrdinals.Empty() {
> > 				fmt.Fprintf(f.Buffer, " [column ordinals: %s]", dep.ColumnOrdinals)
> > ```
> 
> I don't think we need to print both the ordinals and the names. If we have names, we should use them.
> 
> _[pkg/sql/opt/optbuilder/builder.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94XFD21ryUtlhNuUsd:-M94XFD21ryUtlhNuUse:ben42d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/builder.go#L342)):_
> 
> > ```go
> > }
> > 
> > // TrackReferencedColumnForViews is used to add a column to a views dependencies.
> > ```
> 
> [nit] to the view dependencies
> 
> _[pkg/sql/opt/optbuilder/testdata/create_view, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94Xu-v44Lg5YQGZ4cT:-M94Xu-v44Lg5YQGZ4cU:b249avz) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/testdata/create_view#L93)):_
> 
> > ```
> >  └── dependencies
> >       ├── ab [column ordinals: (0)] [column names: [a]]
> >       ├── ab
> > ```
> 
> does this mean we don't depend on any columns? maybe change the formatter to print that explicitly somehow (e.g. `ab [no columns]`).
> 
> _[pkg/sql/opt/optbuilder/testdata/create_view, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94Y9PuCksSLQsuYfnU:-M94Y9Pv5LgbUBI__YdT:b-sk34fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/testdata/create_view#L106)):_
> 
> > ```
> > error (42601): CREATE VIEW specifies 1 column name, but data source has 2 columns
> > 
> > build
> > ```
> 
> Why did you remove this comment?

Addressed these comments.

Also regarding the [no columns] I added this in but theres a small difference where column dependencies aren't tracked when selecting from a view, ie view dependency on another view. I think this makes sense since adding a dependency on the view itself but not the view's columns makes sense.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rytaft)

___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94Y9PuCksSLQsuYfnU:-M955msW9AnplsAoW-lb:b-fmfs6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/testdata/create_view#L106)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Addressed these comments.

Also regarding the [no columns] I added this in but theres a small difference where column dependencies aren't tracked when selecting from a view, ie view dependency on another view. I think this makes sense since adding a dependency on the view itself but not the view's columns makes sense. 
</blockquote></details>

Sounds like something we will need to fix at some point too. We can make `GetColumnNames` return an `ok` flag and we would only show "no columns" if it returned ok=true and empty list.


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rytaft)
> 
> _[pkg/sql/opt/optbuilder/testdata/create_view, line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M94Y9PuCksSLQsuYfnU:-M955msW9AnplsAoW-lb:b-fmfs6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/27540619f7a7651f497d50d758ea62c83fb4f627/pkg/sql/opt/optbuilder/testdata/create_view#L106)):_
> 
> Previously, RichardJCai (Richard Cai) wrote…
> > Addressed these comments.
> > Also regarding the [no columns] I added this in but theres a small difference where column dependencies aren't tracked when selecting from a view, ie view dependency on another view. I think this makes sense since adding a dependency on the view itself but not the view's columns makes sense.
> 
> Sounds like something we will need to fix at some point too. We can make `GetColumnNames` return an `ok` flag and we would only show "no columns" if it returned ok=true and empty list.

I'm not sure we would have to "fix" it right? Since we don't really depend on another view's columns? I don't think we can change a views column set (not 100% sure), but even renaming view dependent columns in Postgres doesn't change the column set of the view. But yeah I can add that check where it doesn't show columns if the dependency is another view.

rytaft reviewed (changes requested):
> Reviewed 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @RichardJCai)

___
*[pkg/sql/opt/view_dependencies.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9IqUg86DziEL4grPdS:-M9IqUg86DziEL4grPdT:bchr97m) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/view_dependencies.go#L31)):*
> ```Go
> 	ColumnOrdinals util.FastIntSet
> 
> 	ColumnIDToOrd map[ColumnID]int
> ```

Still need a comment for this
___
*[pkg/sql/opt/view_dependencies.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9IqnpU0S3eNTCCm4SZ:-M9IqnpV2tOPUijMg6Bi:b-v3t4xk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/view_dependencies.go#L39)):*
> ```Go
> }
> 
> // GetColumnNames returns a sorted list of the name of the column dependencies
> ```

[nit] name -> names
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9IrmIFAhqVfKZpjxph:-M9IrmIFAhqVfKZpjxpi:brebn5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L72)):*
> ```
>  └── dependencies
>       ├── av
>       └── ab [column names: [a]]
> ```

I would simplify this formatting to look like this:
```
└── ab [columns: a]
```
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9Isptb0aFyfO7KpBQn:-M9Isptb0aFyfO7KpBQo:b-wz95n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L93)):*
> ```
>  └── dependencies
>       ├── ab [column names: [a]]
>       ├── ab [no columns]
> ```

Will it be a problem that we have two different dependencies listed for `ab` with different columns?
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 325 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9ItfD_2SEaE30YP0yx:-M9ItfD_2SEaE30YP0yy:bsyfaqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L325)):*
> ```
>  ├── columns: sum:5
>  └── dependencies
>       └── abc [column names: [a b]]
> ```

Where's the dependency on c?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/view_dependencies.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9IqUg86DziEL4grPdS:-M9JbY3hBqIHoy4VHEFI:b34ymx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/view_dependencies.go#L31)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Still need a comment for this
</blockquote></details>

Added this comment.
___
*[pkg/sql/opt/view_dependencies.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9IqnpU0S3eNTCCm4SZ:-M9JbZXX36W4KsKF8oMV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/view_dependencies.go#L39)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] name -> names
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9IrmIFAhqVfKZpjxph:-M9Jbc1pDOQMVIy6RBVx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L72)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I would simplify this formatting to look like this:
```
└── ab [columns: a]
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9Isptb0aFyfO7KpBQn:-M9JXl79DnNvkCfENB5d:b-wkl7da) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L93)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Will it be a problem that we have two different dependencies listed for `ab` with different columns?
</blockquote></details>

I don't think it will be a problem, the same table is used as a data source multiple times where different column sets are used.
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 325 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9ItfD_2SEaE30YP0yx:-M9JXYkG6ASqJNo2aRBp:b-vfd5u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L325)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Where's the dependency on c?
</blockquote></details>

Oops missed this check. Fixed it.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @RichardJCai)

___
*[pkg/sql/opt/memo/expr_format.go, line 557 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9JgztbCKTXXjO2hhhw:-M9JgztbCKTXXjO2hhhx:b-m2727a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8dfb20dc08e2bfdec158762369de22e40398f99/pkg/sql/opt/memo/expr_format.go#L557)):*
> ```Go
> 			colNames, isTable := dep.GetColumnNames()
> 			if len(colNames) > 0 {
> 				fmt.Fprintf(f.Buffer, " [column names:")
> ```

column names -> columns 
(shorter is better given how must is printed in these opt plans)
___
*[pkg/sql/opt/optbuilder/window.go, line 410 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9Jhwkr58i4VXrTBbbE:-M9Jhwks9IxRfO5NK7tv:b7xtrek) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8dfb20dc08e2bfdec158762369de22e40398f99/pkg/sql/opt/optbuilder/window.go#L410)):*
> ```Go
> 			col := outScope.findExistingCol(e, false /* allowSideEffects */)
> 			if col != nil {
> 				b.TrackReferencedColumnForViews(col)
> ```

Do you need to call `b.TrackReferencedColumnForViews` every time the function `findExistingCol` is used? In that case, maybe you should move this call inside that function. I think that would be a bit less error-prone, and also help avoid future errors.
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9Isptb0aFyfO7KpBQn:-M9Jf89u62-Aik_WmHfs:bqn23zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L93)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I don't think it will be a problem, the same table is used as a data source multiple times where different column sets are used.
</blockquote></details>

But have you confirmed (maybe with some logic tests) that those dependencies are respected when you try to drop those columns? Looking at the code in `opt_exec_factory.go` it looks like the dependencies are correctly combined for each table, but it would be good to add some tests for future-proofing.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/expr_format.go, line 557 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9JgztbCKTXXjO2hhhw:-M9JzKAg-VU7QyCKQXUS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8dfb20dc08e2bfdec158762369de22e40398f99/pkg/sql/opt/memo/expr_format.go#L557)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

column names -> columns 
(shorter is better given how must is printed in these opt plans)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/window.go, line 410 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9Jhwkr58i4VXrTBbbE:-M9JzTU71rjghX11osLy:b3jb6nq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8dfb20dc08e2bfdec158762369de22e40398f99/pkg/sql/opt/optbuilder/window.go#L410)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you need to call `b.TrackReferencedColumnForViews` every time the function `findExistingCol` is used? In that case, maybe you should move this call inside that function. I think that would be a bit less error-prone, and also help avoid future errors.
</blockquote></details>

Good call,  I checked and this made sense to me, I moved the TrackReferencedColumnsForViews call into findExistingCol and added a comment about findExistingCol will add a view dep if a column is found.
___
*[pkg/sql/opt/optbuilder/testdata/create_view, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9Isptb0aFyfO7KpBQn:-M9JzMsq2xsGPvyNUxuZ:bvdw36m) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd82d5ff21c355469a11295e17a50e9df46d60be/pkg/sql/opt/optbuilder/testdata/create_view#L93)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

But have you confirmed (maybe with some logic tests) that those dependencies are respected when you try to drop those columns? Looking at the code in `opt_exec_factory.go` it looks like the dependencies are correctly combined for each table, but it would be good to add some tests for future-proofing.
</blockquote></details>

Sounds good, I added some logic tests for the case where the table is used as a data source multiple times with different column sources


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 8 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @RichardJCai)

___
*[pkg/sql/logictest/testdata/logic_test/views, line 693 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9K3UI75MYt4-S2134I:-M9K3UI75MYt4-S2134J:b-5niwsc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f8f31fda6b6a86d3510cadcc0c341c4a9676a35/pkg/sql/logictest/testdata/logic_test/views#L693)):*
> ```
> 
> # Ensure a table that is referenced multiple times with different column sets
> # depends on the correct columns. Depended on columns should not be droppable.
> ```

[nit] I'd change this to "Ensure a view that contains a table that is referenced multiple times...."
___
*[pkg/sql/opt/optbuilder/orderby.go, line 106 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9K3rJ37Ipl6Ke8EKIi:-M9K3rJ37Ipl6Ke8EKIj:b-tm7dfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f8f31fda6b6a86d3510cadcc0c341c4a9676a35/pkg/sql/opt/optbuilder/orderby.go#L106)):*
> ```Go
> 	// projection.
> 	if col := projectionsScope.findExistingCol(
> 		b,
> ```

you shouldn't need to pass in `b` -- scope already contains `builder` as a data member.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/views, line 693 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9K3UI75MYt4-S2134I:-M9KKM1y2gZ_ZKUPmtYG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f8f31fda6b6a86d3510cadcc0c341c4a9676a35/pkg/sql/logictest/testdata/logic_test/views#L693)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] I'd change this to "Ensure a view that contains a table that is referenced multiple times...."
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/orderby.go, line 106 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9K3rJ37Ipl6Ke8EKIi:-M9KKGrr4V3qtiEec4GT:b-wtr871) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f8f31fda6b6a86d3510cadcc0c341c4a9676a35/pkg/sql/opt/optbuilder/orderby.go#L106)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you shouldn't need to pass in `b` -- scope already contains `builder` as a data member.
</blockquote></details>

You're right, updated.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for working through all these changes!

> Reviewed 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-:-M9KcuCD2sFjygnh3EaN:b-9wb0rh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @RichardJCai)

___
*[pkg/sql/opt/optbuilder/builder.go, line 345 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9KdJSG8aH5c9EY8Vjt:-M9KdJSG8aH5c9EY8Vju:b7r8tew) ([raw file](https://github.com/cockroachdb/cockroach/blob/c21d18c4cb2e3cc970adc58a19c9caabed165102/pkg/sql/opt/optbuilder/builder.go#L345)):*
> ```Go
> // dependencies. This should be called whenever a column reference is made in a
> // view query.
> func (b *Builder) TrackReferencedColumnForViews(col *scopeColumn) {
> ```

[nit] does this function need to be exported? Seems like we're not using it outside of optbuilder, right?


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/optbuilder/builder.go, line 345 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49872#-M9KdJSG8aH5c9EY8Vjt:-M9KyOR_EQvd3k5tf4U5:bva88ou) ([raw file](https://github.com/cockroachdb/cockroach/blob/c21d18c4cb2e3cc970adc58a19c9caabed165102/pkg/sql/opt/optbuilder/builder.go#L345)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] does this function need to be exported? Seems like we're not using it outside of optbuilder, right?
</blockquote></details>

Good point, updated to not export it.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49872)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RichardJCai commented:
bors r=rytaft

RichardJCai commented:
Thanks for the reviews!
